File: neutralino.js/src/core/os.js
// MIT License

// Copyright (c) 2018 Neutralinojs

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

let $ = require('../lib/minAjax.js');

let runCommand = function(cmd, s, e) {
    $.ajax({
        url : '/os/runCommand',
        type : 'POST',
        data : {
          command : cmd
        },
        done : function(data){
            s(data);
        },
        problem : function (error) {
            e(error);
        }

    });

};

let getEnvar = function(v, s, e) {
    $.ajax({
        url : '/os/getEnvar',
        type : 'POST',
        data : {
          name : v
        },
        done : function(data){
            s(data);
        },
        problem : function(error) {
            e(error);
        }

    });

};


let dialogOpen = function(options, s, e) {
    $.ajax({
        url : '/os/dialogOpen',
        type : 'POST',
        data : {
          title : options["title"],
          isDirectoryMode : options["isDirectoryMode"]
        },
        done : function(data){
            s(data);
        },
        problem : function(error) {
            e(error);
        }

    });

};


let dialogSave = function(t, s, e) {
    $.ajax({
        url : '/os/dialogSave',
        type : 'POST',
        data : {
          title : t
        },
        done : function(data){
            s(data);
        },
        problem : function(error) {
            e(error);
        }

    });

};

let showNotification = function(options, s, e) {
    $.ajax({
        url : '/os/showNotification',
        type : 'POST',
        data : options,
        done : function(data){
            s(data);
        },
        problem : function(error) {
            e(error);
        }

    });

};

let showMessageBox = function(options, s, e) {
    $.ajax({
        url : '/os/showMessageBox',
        type : 'POST',
        data : options,
        done : function(data){
            s(data);
        },
        problem : function(error) {
            e(error);
        }

    });

};

module.exports = {
    runCommand : runCommand,
    getEnvar : getEnvar,
    dialogOpen : dialogOpen,
    dialogSave : dialogSave,
    showNotification: showNotification,
    showMessageBox: showMessageBox
}

File: core-windows/src/core/os/os.cpp
// MIT License

// Copyright (c) 2018 Neutralinojs

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include <iostream>
#include <fstream>
#include "../../../lib/json/json.hpp"
#include <windows.h>
#include <shlobj.h>
#include <shobjidl.h>
#include <stdlib.h>
#include <cstdio>
#include <iostream>
#include <memory>
#include <stdexcept>
#include <string>
#include <array>

#include "../../functions.h"

using namespace std;
using json = nlohmann::json;

namespace os {

    string runCommand(string jso) {
        json input;
        json output;
        try {
            input = json::parse(jso);
        }
        catch(exception e){
            output["error"] = "JSON parse error is occurred!";
            return output.dump();
        }
        
        string command = "cmd /c " + input["command"].get<std::string>();
        output["stdout"] = functions::execCommand(command);
        return output.dump();
    }

    string getEnvar(string jso) {
        json input;
        json output;
        try {
            input = json::parse(jso);
        }
        catch(exception e){
            output["error"] = "JSON parse error is occurred!";
            return output.dump();
        }
        string i = input["name"];
        char *o;
        o = getenv(i.c_str());
        if(o == NULL) {
            output["error"] =  i + " is not defined";
        }
        else {
            output["value"] = o;
        }
        return output.dump();
       
        
    }


    string dialogOpen(string jso) {
        json input;
        json output;
        try {
            input = json::parse(jso);
        }
        catch(exception e){
            output["error"] = "JSON parse error is occurred!";
            return output.dump();
        }
        string title = input["title"];
        if(!input["isDirectoryMode"].is_null() && input["isDirectoryMode"].get<bool>()) {
            TCHAR szDir[MAX_PATH];
            BROWSEINFO bInfo;
            ZeroMemory(&bInfo, sizeof(bInfo));
            bInfo.hwndOwner = NULL;
            bInfo.pidlRoot = NULL; 
            bInfo.pszDisplayName = szDir;
            bInfo.lpszTitle = const_cast<char *>(title.c_str());
            bInfo.ulFlags = 0 ;
            bInfo.lpfn = NULL;
            bInfo.lParam = 0;
            bInfo.iImage = -1;

            LPITEMIDLIST lpItem = SHBrowseForFolder( &bInfo);
            if( lpItem != NULL ) {
                SHGetPathFromIDList(lpItem, szDir );
                output["file"] = szDir;
            }
            else {
                output["file"] = "";
            }
        }
        else {
            OPENFILENAME ofn;
            TCHAR szFile[MAX_PATH] = { 0 }; 
            ZeroMemory(&ofn, sizeof(ofn));
            ofn.lpstrTitle = const_cast<char *>(title.c_str());
            ofn.lStructSize = sizeof(ofn);
            ofn.lpstrFile = szFile;
            ofn.nMaxFile = sizeof(szFile);
            ofn.nFilterIndex = 1;
            ofn.lpstrFileTitle = NULL;
            ofn.nMaxFileTitle = 0;
            ofn.lpstrInitialDir = NULL;
            ofn.Flags = OFN_PATHMUSTEXIST | OFN_FILEMUSTEXIST;

            if (GetOpenFileName(&ofn)) {
                output["file"] = ofn.lpstrFile;
            }
            else {
                output["file"] = "";
            }
        }

        return output.dump();
       
        
    }


    string dialogSave(string jso) {
        json input;
        json output;
        try {
            input = json::parse(jso);
        }
        catch(exception e){
            output["error"] = "JSON parse error is occurred!";
            return output.dump();
        }
        string title = input["title"];
        OPENFILENAME ofn;
        TCHAR szFile[260] = { 0 }; 

        ZeroMemory(&ofn, sizeof(ofn));
        ofn.lpstrTitle = const_cast<char *>(title.c_str());
        ofn.lStructSize = sizeof(ofn);
        ofn.lpstrFile = szFile;
        ofn.nMaxFile = sizeof(szFile);
        ofn.nFilterIndex = 1;
        ofn.lpstrFileTitle = NULL;
        ofn.nMaxFileTitle = 0;
        ofn.lpstrInitialDir = NULL;
        ofn.Flags = OFN_PATHMUSTEXIST;

        if (GetSaveFileName(&ofn)) {
            output["file"] = ofn.lpstrFile;
        }
        else {
            output["file"] = "";
        }

        return output.dump();
       
        
    }

    string showNotification(string jso) {
        json input;
        json output;
        try {
            input = json::parse(jso);
        }
        catch(exception e){
            output["error"] = "JSON parse error is occurred!";
            return output.dump();
        }
        string command = "powershell -Command \"& {[reflection.assembly]::loadwithpartialname('System.Windows.Forms');"
                        "[reflection.assembly]::loadwithpartialname('System.Drawing');"
                        "$notify = new-object system.windows.forms.notifyicon;"
                        "$notify.icon = [System.Drawing.SystemIcons]::Information;"
                        "$notify.visible = $true;"
                        "$notify.showballoontip(0 ,'"+ input["summary"].get<string>() + "','" + input["body"].get<string>() + "',[system.windows.forms.tooltipicon]::None)}\"";
        
        string commandOutput = functions::execCommand(command);

        if(commandOutput.find("'powershell'") != string::npos)
            output["message"] = "Notification is pushed to the system";
        else
            output["error"] = "An error thrown while sending the notification";
        return output.dump();
        
    }

    
}
